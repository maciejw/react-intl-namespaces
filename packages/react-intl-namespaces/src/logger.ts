export interface Logger {
  error(...params: any[]): void;
  warn(...params: any[]): void;
  info(...params: any[]): void;
  debug(...params: any[]): void;
}
export const dummyConsole = {
  debug: () => {
    return;
  },
  error: () => {
    return;
  },
  info: () => {
    return;
  },
  warn: () => {
    return;
  },
};
// tslint:disable:no-bitwise
// prettier-ignore
export enum LogLevel {
  none  =   0,
  error =   1 << 0,
  warn  =  (1 << 1) | error,
  info  =  (1 << 2) | warn,
  debug =  (1 << 3) | info,
}

let { error, info, warn, debug } = window.console || dummyConsole;

debug = debug.bind(window.console);
error = error.bind(window.console);
info = info.bind(window.console);
warn = warn.bind(window.console);

class ConsoleLogger implements Logger {
  constructor(private name?: string) {}

  public error(...params: any[]): void {
    if ((Logger.logLevel & LogLevel.error) === LogLevel.error) {
      this.log(error, ...params);
    }
  }
  public warn(...params: any[]): void {
    if ((Logger.logLevel & LogLevel.warn) === LogLevel.warn) {
      this.log(warn, ...params);
    }
  }
  public info(...params: any[]): void {
    if ((Logger.logLevel & LogLevel.info) === LogLevel.info) {
      this.log(info, ...params);
    }
  }
  public debug(...params: any[]): void {
    if ((Logger.logLevel & LogLevel.debug) === LogLevel.debug) {
      this.log(debug, ...params);
    }
  }
  private log(log: (...params: any[]) => void, ...params: any[]) {
    const name = this.getLoggerName();
    if (name) {
      log(...[name, ...params]);
    } else {
      log(...params);
    }
  }
  private getLoggerName() {
    if (this.name) {
      return `[${this.name}]`;
    }
    return;
  }
}

namespace Logger {
  export let logger: Logger = new ConsoleLogger();
  export let logLevel = LogLevel.none;
}

export function setLogLevel(level: LogLevel) {
  Logger.logLevel = level;
}

export function mockConsole(console: Console) {
  info = console.info;
  debug = console.debug;
  warn = console.warn;
  error = console.error;
}

function createLogger(name: string): Logger {
  return new ConsoleLogger(name);
}
const { logger } = Logger;

export { logger, createLogger };
