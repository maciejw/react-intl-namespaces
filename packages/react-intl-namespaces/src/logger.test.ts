import * as Logger from './logger';

const { configureLogger, Level, setConsole } = Logger;
const ConsoleMock = jest.fn<Console>(() => ({
  debug: jest.fn(),
  error: jest.fn(),
  info: jest.fn(),
}));

describe('Logger', () => {
  it('should should not log by default', () => {
    const consoleMock = new ConsoleMock();
    const { logger } = Logger;
    setConsole(consoleMock);

    logger.debug('debug');
    logger.info('info');
    logger.error('error');

    expect(consoleMock).toMatchSnapshot();
  });
  it('should log error only to console when configured accordingly', () => {
    const consoleMock = new ConsoleMock();

    setConsole(consoleMock);
    configureLogger(Level.error);
    const { logger } = Logger;

    logger.debug('debug');
    logger.info('info');
    logger.error('error');

    expect(consoleMock).toMatchSnapshot();
  });
  it('should log info only to console when configured accordingly', () => {
    const consoleMock = new ConsoleMock();
    setConsole(consoleMock);
    configureLogger(Level.info);

    const { logger } = Logger;

    logger.debug('debug');
    logger.info('info');
    logger.error('error');

    expect(consoleMock).toMatchSnapshot();
  });
  it('should log debug only to console when configured accordingly', () => {
    const consoleMock = new ConsoleMock();

    setConsole(consoleMock);
    configureLogger(Level.debug);
    const { logger } = Logger;

    logger.debug('debug');
    logger.info('info');
    logger.error('error');

    expect(consoleMock).toMatchSnapshot();
  });
});
