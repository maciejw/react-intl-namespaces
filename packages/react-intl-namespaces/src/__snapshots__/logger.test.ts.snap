// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logger should log debug only to console when configured accordingly 1`] = `
Object {
  "debug": [MockFunction] {
    "calls": Array [
      Array [
        "debug",
      ],
    ],
  },
  "error": [MockFunction] {
    "calls": Array [
      Array [
        "error",
      ],
    ],
  },
  "info": [MockFunction] {
    "calls": Array [
      Array [
        "info",
      ],
    ],
  },
}
`;

exports[`Logger should log error only to console when configured accordingly 1`] = `
Object {
  "debug": [MockFunction],
  "error": [MockFunction] {
    "calls": Array [
      Array [
        "error",
      ],
    ],
  },
  "info": [MockFunction],
}
`;

exports[`Logger should log info only to console when configured accordingly 1`] = `
Object {
  "debug": [MockFunction],
  "error": [MockFunction] {
    "calls": Array [
      Array [
        "error",
      ],
    ],
  },
  "info": [MockFunction] {
    "calls": Array [
      Array [
        "info",
      ],
    ],
  },
}
`;

exports[`Logger should should not log by default 1`] = `
Object {
  "debug": [MockFunction],
  "error": [MockFunction],
  "info": [MockFunction],
}
`;
